[quant] ConvTranspose1d / ConvTranspose2d

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43199 [quant] conv_transpose graph patterns
* **#40371 [quant] ConvTranspose1d / ConvTranspose2d**
* #40370 [quant] conv_transpose1d / conv_transpose2d

Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

Commit:
[quant] ConvTranspose1d / ConvTranspose2d

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

z-a-f commented:
FYI: https://github.com/pytorch/pytorch/pull/36790

vkuzo reviewed (commented):
would be good to add test coverage, we can review more in detail once there is more evidence on the memory corruption bug

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

kimishpatel commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -45,8 +45,12 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
         self.padding_mode = padding_mode
         # Initialize as NCHW. set_weight will internally transpose to NHWC.
+        if self.transposed:
+            weight_shape = [in_channels, out_channels // self.groups]
```
I suspect something wrong here. Are you sure dimensions are supposed to be like this. I see below that `bias_float`'s size is out_channels. It should also depend on transposed param? So here : https://github.com/pytorch/pytorch/pull/40370/files#diff-d0713fbda560de6aecef93e4c3d973a5R561, you are providing #of output channels differently than what your weight is telling you.
But more importantly what is going on with I dont understand why in deconv weigth's #of output channels is equal to in_channel? I would have thought it should not change right? Your deconv op's weight should tell you how many output channels you want to produce and that should not depend on # of input channels. No? Only thing different here I imagine is output shape calculation.

z-a-f replied:
bias size should always be equal to the number of output channels. The shape of the weight matrix in convolution and transposed convolution is different:

- convolution weight shape: [out_channels, in_channels / groups, k0, k1, k2, ...]
- convolution transpose weight shape: [in_channels, out_channels / groups, k0, k1, k2, ...]

Can you elaborate what you mean when you say "deconv weights #of output channels equals to in_channels"?

z-a-f replied:
Thanks Kimish for finding the bug!

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

z-a-f commented:
> would be good to add test coverage, we can review more in detail once there is more evidence on the memory corruption bug

The unittests are there, part of the previous PRs

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

vkuzo commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -508,3 +509,194 @@ def _input_padding(self, kernel_size, dilation, padding):
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, output_padding=0, groups=1, bias=True,
+                 dilation=1, padding_mode='zeros'):
+        kernel_size = _pair(kernel_size)
+        stride = _pair(stride)
+        padding = _pair(padding)
+        dilation = _pair(dilation)
+        output_padding = _pair(output_padding)
```
can these live on the base class?

z-a-f replied:
I don't think so -- these are supposed to be different for every spatial dim (`_single`, `_pair`, `_triple`, etc.). The reason we are reusing because it's a "hack" to utilize the 2d to do the 1d computation. However, in the future this will change. If we put it in the base class, this might propagate through. You can take a look at the `conv.py` in the non-quantized space -- that's the pattern I am aiming for.

vkuzo commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -508,3 +509,194 @@ def _input_padding(self, kernel_size, dilation, padding):
+        w, b = torch.ops.quantized.conv_transpose1d_unpack(self._packed_params)
+        return w, b
+
+    def weight(self):
+        (w, _) = self._weight_bias()
+        return w
+
+    def bias(self):
+        (_, b) = self._weight_bias()
+        return b
```
and these on the base class?

z-a-f replied:
Sure will do -- I guess we should add tests for the ConvXd modules as well.

vkuzo reviewed (commented):
looks good.  should we add a test?

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

neuromorpheus commented:
Hi,

When is this feature expected to be available in pytorch?

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

z-a-f commented:
> Hi,
> 
> When is this feature expected to be available in pytorch?

Hopefully soon. You mind I ask where you use it?

neuromorpheus commented:
I want to use it for upsampling quantized tensors. I know I can use
interpolate, but I wanted to compare timings with a transposed convolution.

On Tue, Aug 25, 2020, 5:55 AM Zafar <notifications@github.com> wrote:

> Hi,
>
> When is this feature expected to be available in pytorch?
>
> Hopefully soon. You mind I ask where you use it?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/40371#issuecomment-679927336>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOQJASZY5NY67PEEJZBEUUDSCODCHANCNFSM4OEMUGTA>
> .
>

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

vkuzo reviewed (approved):
lg as long as CI passes.  Can we include the context and the test plan in the PR summary before landing, since this is non-trivial?

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]

Commit:
Update on "[quant] ConvTranspose1d / ConvTranspose2d"


Differential Revision: [D22158981](https://our.internmc.facebook.com/intern/diff/D22158981)

[ghstack-poisoned]
